File: Hashtable_Console/src/main/java/tn/usousse/eniso/ia1/stage/example/service/Service.java
package tn.usousse.eniso.ia1.stage.example.service;

import tn.usousse.eniso.ia1.stage.example.model.Node;
import tn.usousse.eniso.ia1.stage.example.model.Table;

import java.lang.Math;
public class Service {
    private final Table table;

    public Service(Table t){

        this.table = t;
    }

    public boolean exists(String value, Table table){
        int i = hash(value);
        Node n = table.getNodes()[i];
        while ( n != null){
            if (n.getValue().equals(value)){
                return true;
            }
            n = n.getNext();
        }
        return false;
    }
    public int hash(String value){
        int i =0;
        for(int j = 0; j<value.length(); j++){
            i+=(int) value.charAt(j);
        }
        return Math.abs(i%3);
    }


    public boolean add(String value) {
        if (exists(value, this.table)) {
            return false;
        }

        int i = hash(value);
        Node n = new Node(value);

        if (this.table.getNodes()[i] == null) {
            this.table.getNodes()[i] = n;

        }

        else {
            Node currentNode = this.table.getNodes()[i];
            while (currentNode.getNext() != null) {
                currentNode = currentNode.getNext();
            }
            currentNode.setNext(n);
        }

        return true;
    }
    public boolean remove(String value)  {
        int i = hash(value);
        if (table.getNodes()[i] == null){
            return false;
        }
        Node p = table.getNodes()[i];
        Node q = null;
        while(p!= null){
            if(value.equals(p.getValue())){
                if(q == null){
                    table.getNodes()[i] = p.getNext();
                }
                else {
                    q.setNext(p.getNext());
                }
                return true;
            }
            q = p;
            p = p.getNext();
        }

        return false;
    }
    public Table list(){
        Table list = new Table(table.getNodes().length);
        for(int i = 0; i<table.getNodes().length; i++){
            Node linkedList = table.getNodes()[i];
            list.getNodes()[i] = linkedList;
        }
        return list;
    }
}

File: Hashtable_Angular/src/app/service/Service.ts
import {Table} from "../dto/Table";
import {Node} from  "../dto/Node"
export class Service{
  private _table: Table;
  constructor(t: Table) {
    this._table = t;
  }

  get table(): Table {
    return this._table;
  }

  set table(value: Table) {
    this._table = value;
  }
  public exists(value: string, table: Table): boolean {
    const index: number = this.hash(value);
    let node: Node | null = table.nodes[index];

    while (node !== null) {
      if (node.value === value) {
        return true;
      }
      node = node.next;
    }

    return false;
  }
  public hash(value: string): number {
    let index: number = 0;

    for (let j = 0; j < value.length; j++) {
      index += value.charCodeAt(j);
      index = index * 31;
    }

    return Math.abs(index % this._table.nodes.length);
  }
  public add(value: string): boolean{
    if(this.exists(value,this._table)){
      return false;
    }
    const index: number = this.hash(value);
    const node: Node = new Node(value);
    if (this.table.nodes[index] == null) {
      this.table.nodes[index] = node;
    }
    else {
      let currentNode: Node |null = this.table.nodes[index] as Node;
      while (currentNode.next != null) {
        currentNode = currentNode.next;
      }
      currentNode.next = node;
    }
    return true
  }
  public remove(value: string): boolean {
    const i: number = this.hash(value);
    if (this._table.nodes[i] == null) {
      return false;
    }
    let p: Node | null = this.table.nodes[i];
    let q: Node | null = null;
    while (p !== null) {
      if (value === p.value) {
        if (q === null) {
          this._table.nodes[i] = <Node>p.next;
        } else {
          q.next = p.next;
        }
        return true;
      }
      q = p;
      p = p.next;
    }

    return false;
  }

}
